pkg/operator/crdstate.go

```diff
@@ -22,34 +22,47 @@ import (
 	"k8s.io/client-go/tools/cache"
 )
 
-var errInvalidCond = fmt.Errorf("condition needs both 'Type' and 'Status' fields set")
+var (
+	errInvalidCond     = fmt.Errorf("condition needs both 'Type' and 'Status' fields set")
+	errUnsupportedType = fmt.Errorf("unsupported type for CRD state")
+)
 
 // CRDStatusState maintains state of the statuses of CRDs the operator manages.
 type CRDStatusState struct {
-	podmons map[string]*pmState
+	pMons map[string]*pmState
+	cMons map[string]*pmState
 	// Primarily used for testing.
 	now func() metav1.Time
 }
 
 type pmState struct {
 	pm          *monitoringv1alpha1.PodMonitoring
+	cm          *monitoringv1alpha1.ClusterPodMonitoring
 	needsUpdate bool
 	conds       map[monitoringv1alpha1.MonitoringConditionType]*monitoringv1alpha1.MonitoringCondition
 }
 
-func newPMState(pm *monitoringv1alpha1.PodMonitoring, now metav1.Time) *pmState {
+func newPMState(pm *monitoringv1alpha1.PodMonitoring, cm *monitoringv1alpha1.ClusterPodMonitoring, now metav1.Time) *pmState {
```

Comment:
I refactored a bit.

Since these are all instance methods, I didn't see returning an extra slice in the various function signatures as a huge win. Taking advantage of the `collectionReconciler`'s state and using a simple resource object slice there felt cleaner, so that's what I did.

Happy to discuss more.